server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java

```diff
@@ -419,6 +502,98 @@ private Set<SegmentId> filterSegmentWithCachedSchema(Set<SegmentId> segmentIds)
     return cachedSegments;
   }
 
+  @VisibleForTesting
+  protected void coldDatasourceSchemaExec()
+  {
+    Collection<ImmutableDruidDataSource> immutableDataSources =
+        sqlSegmentsMetadataManager.getImmutableDataSourcesWithAllUsedSegments();
+
+    final Map<String, ColumnType> columnTypes = new LinkedHashMap<>();
+
+    Set<String> dataSources = new HashSet<>();
+
+    for (ImmutableDruidDataSource dataSource : immutableDataSources) {
+      String dataSourceName = dataSource.getName();
+      dataSources.add(dataSourceName);
+      Collection<DataSegment> dataSegments = dataSource.getSegments();
+
+      for (DataSegment segment : dataSegments) {
+        Integer replicationFactor = segmentReplicationStatusManager.getReplicationFactor(segment.getId());
+        if (replicationFactor != null && replicationFactor != 0) {
+          continue;
+        }
+        Optional<SchemaPayloadPlus> optionalSchema = segmentSchemaCache.getSchemaForSegment(segment.getId());
+        if (optionalSchema.isPresent()) {
+          RowSignature rowSignature = optionalSchema.get().getSchemaPayload().getRowSignature();
+          for (String column : rowSignature.getColumnNames()) {
+            final ColumnType columnType =
+                rowSignature.getColumnType(column)
+                            .orElseThrow(() -> new ISE("Encountered null type for column [%s]", column));
+
+            columnTypes.compute(column, (c, existingType) -> columnTypeMergePolicy.merge(existingType, columnType));
+          }
+        }
+      }
+
+      final RowSignature.Builder builder = RowSignature.builder();
+      columnTypes.forEach(builder::add);
+
+      RowSignature coldSignature = builder.build();
+
+      log.debug("[%s] signature from cold segments is [%s]", dataSourceName, coldSignature);
+
+      coldSchemaTable.put(dataSourceName, new DataSourceInformation(dataSourceName, coldSignature));
+
+      // update tables map with merged schema, if signature doesn't exist we do not add entry in this table
+      // schema for entirely cold datasource is maintained separately
+      tables.computeIfPresent(
+          dataSourceName,
+          (ds, info) -> {
+            RowSignature mergedSignature = mergeHotAndColdSchema(info.getRowSignature(), coldSignature);
+
+            if (!info.getRowSignature().equals(mergedSignature)) {
+              log.info(
+                  "[%s] has new merged signature: %s. hot signature [%s], cold signature [%s].",
+                  ds, mergedSignature, info.getRowSignature(), coldSignature
+              );
+            } else {
+              log.debug("[%s] merged signature is unchanged.", ds);
+            }
+
+            return new DataSourceInformation(ds, mergedSignature);
+          }
+      );
+    }
+
+    // remove any stale datasource from the map
+    coldSchemaTable.keySet().retainAll(dataSources);
```

Comment:
Do we have a test case for this ?

Reply:
Yes, in `CoordinatorSegmentMetadataCacheTest#testColdDatasourceSchema_verifyStaleDatasourceRemoved`.